File: node/src/bulk.js
/**
 * VedaDB Node.js Driver - Bulk Inserter + Pipeline
 *
 * High-performance batch insertions with automatic batching,
 * pipeline execution, and progress tracking.
 */

'use strict';

const { EventEmitter } = require('events');
const { BulkError } = require('./errors');

/**
 * Escape a value for SQL.
 * @param {*} v
 * @returns {string}
 */
function esc(v) {
  if (v === null || v === undefined) return 'NULL';
  if (typeof v === 'boolean') return v ? 'TRUE' : 'FALSE';
  if (typeof v === 'number') return String(v);
  if (typeof v === 'string') return `'${v.replace(/'/g, "''")}'`;
  if (v instanceof Date) return `'${v.toISOString()}'`;
  if (typeof v === 'object') return `'${JSON.stringify(v).replace(/'/g, "''")}'`;
  return String(v);
}

/**
 * Bulk inserter for efficient batch INSERT operations.
 * Accumulates rows and flushes when batch size is reached.
 */
class BulkInserter extends EventEmitter {
  /**
   * @param {Object} client - VedaDB client
   * @param {string} table - Target table name
   * @param {number} [batchSize=1000] - Rows per batch
   * @param {Object} [options]
   * @param {boolean} [options.usePipeline=true] - Use pipeline for batches
   * @param {boolean} [options.ignoreDuplicates=false] - Use INSERT IGNORE
   * @param {number} [options.flushIntervalMs=0] - Auto-flush interval (0 = disabled)
   */
  constructor(client, table, batchSize = 1000, options = {}) {
    super();
    this._client = client;
    this._table = table;
    this._batchSize = batchSize;
    this._usePipeline = options.usePipeline !== false;
    this._ignoreDuplicates = options.ignoreDuplicates || false;
    this._flushIntervalMs = options.flushIntervalMs || 0;
    this._rows = [];
    this._total = 0;
    this._flushed = 0;
    this._columns = null;
    this._timer = null;

    if (this._flushIntervalMs > 0) {
      this._timer = setInterval(() => {
        if (this._rows.length > 0) this.flush().catch(() => {});
      }, this._flushIntervalMs);
    }
  }

  /** Number of rows currently buffered. */
  get buffered() {
    return this._rows.length;
  }

  /** Total rows inserted so far. */
  get totalInserted() {
    return this._flushed;
  }

  /** Stats snapshot. */
  get stats() {
    return {
      table: this._table,
      buffered: this._rows.length,
      batchSize: this._batchSize,
      totalInserted: this._flushed,
      pending: this._total - this._flushed,
    };
  }

  /**
   * Add a row to the buffer.
   * Flushes automatically when batch size is reached.
   *
   * @param {Object} row - { column: value }
   * @returns {Promise<number|undefined>} Rows flushed if batch was full
   */
  async add(row) {
    if (!this._columns && row) {
      this._columns = Object.keys(row);
    }

    this._rows.push(row);
    this._total++;

    if (this._rows.length >= this._batchSize) {
      return this.flush();
    }
  }

  /**
   * Add multiple rows at once.
   * @param {Object[]} rows
   * @returns {Promise<number>} Total rows flushed
   */
  async addMany(rows) {
    let flushed = 0;
    for (const row of rows) {
      const result = await this.add(row);
      if (result) flushed += result;
    }
    return flushed;
  }

  /**
   * Manually flush the current buffer.
   * @returns {Promise<number>} Number of rows flushed
   */
  async flush() {
    if (this._rows.length === 0) return 0;

    const batch = this._rows.splice(0, this._rows.length);
    const start = Date.now();

    try {
      let inserted;
      if (this._usePipeline && batch.length > 1) {
        inserted = await this._pipelineInsert(batch);
      } else {
        inserted = await this._singleInsert(batch);
      }

      this._flushed += inserted;
      this.emit('flush', { rows: inserted, duration: Date.now() - start, table: this._table });
      return inserted;
    } catch (err) {
      this.emit('error', { error: err, batchSize: batch.length });
      throw new BulkError(`Bulk insert failed for ${this._table}: ${err.message}`,
        batch.map((_, i) => ({ index: i, error: err })));
    }
  }

  /**
   * Flush remaining rows and close the inserter.
   * @returns {Promise<number>} Total rows flushed in final batch
   */
  async close() {
    if (this._timer) {
      clearInterval(this._timer);
      this._timer = null;
    }
    const result = await this.flush();
    this.emit('close', { totalInserted: this._flushed });
    return result;
  }

  /** Destroy without flushing. */
  destroy() {
    if (this._timer) {
      clearInterval(this._timer);
      this._timer = null;
    }
    this._rows = [];
    this.emit('destroy');
  }

  // -- internal -------------------------------------------------------------

  async _singleInsert(rows) {
    if (!this._columns) return 0;
    const cols = this._columns.join(', ');
    const values = rows.map(row =>
      '(' + this._columns.map(c => esc(row[c])).join(', ') + ')'
    ).join(', ');
    const ignore = this._ignoreDuplicates ? ' IGNORE' : '';
    const sql = `INSERT${ignore} INTO ${this._table} (${cols}) VALUES ${values};`;
    const result = await this._client.query(sql);
    return result.rowCount || rows.length;
  }

  async _pipelineInsert(rows) {
    if (!this._columns) return 0;
    const cols = this._columns.join(', ');
    const ignore = this._ignoreDuplicates ? ' IGNORE' : '';
    const queries = rows.map(row => {
      const vals = this._columns.map(c => esc(row[c])).join(', ');
      return `INSERT${ignore} INTO ${this._table} (${cols}) VALUES (${vals});`;
    });

    const results = await this._client.pipeline(queries);
    return results.length;
  }
}

/**
 * Pipeline builder for batching multiple operations.
 */
class Pipeline extends EventEmitter {
  /**
   * @param {Object} client - VedaDB client
   */
  constructor(client) {
    super();
    this._client = client;
    this._queries = [];
  }

  /** Number of queued queries. */
  get length() {
    return this._queries.length;
  }

  /**
   * Add a query to the pipeline.
   * @param {string} sql
   * @returns {Pipeline}
   */
  query(sql) {
    this._queries.push(sql);
    return this;
  }

  /**
   * Add an INSERT to the pipeline.
   * @param {string} table
   * @param {Object} data
   * @returns {Pipeline}
   */
  insert(table, data) {
    const cols = Object.keys(data);
    const vals = cols.map(c => esc(data[c])).join(', ');
    this._queries.push(`INSERT INTO ${table} (${cols.join(', ')}) VALUES (${vals});`);
    return this;
  }

  /**
   * Add an UPDATE to the pipeline.
   * @param {string} table
   * @param {Object} data - { column: value }
   * @param {Object} where - { column: value }
   * @returns {Pipeline}
   */
  update(table, data, where) {
    const setClause = Object.entries(data)
      .map(([k, v]) => `${k} = ${esc(v)}`)
      .join(', ');
    const whereClause = Object.entries(where || {})
      .map(([k, v]) => {
        if (v === null) return `${k} IS NULL`;
        return `${k} = ${esc(v)}`;
      })
      .join(' AND ');
    let sql = `UPDATE ${table} SET ${setClause}`;
    if (whereClause) sql += ` WHERE ${whereClause}`;
    sql += ';';
    this._queries.push(sql);
    return this;
  }

  /**
   * Add a DELETE to the pipeline.
   * @param {string} table
   * @param {Object} where
   * @returns {Pipeline}
   */
  delete(table, where) {
    const whereClause = Object.entries(where || {})
      .map(([k, v]) => {
        if (v === null) return `${k} IS NULL`;
        return `${k} = ${esc(v)}`;
      })
      .join(' AND ');
    let sql = `DELETE FROM ${table}`;
    if (whereClause) sql += ` WHERE ${whereClause}`;
    sql += ';';
    this._queries.push(sql);
    return this;
  }

  /**
   * Execute all queued queries as a pipeline.
   * @returns {Promise<Result[]>}
   */
  async execute() {
    if (this._queries.length === 0) return [];
    const start = Date.now();
    const queries = this._queries.splice(0);
    const results = await this._client.pipeline(queries);
    this.emit('execute', { count: queries.length, duration: Date.now() - start });
    return results;
  }

  /** Clear queued queries without executing. */
  clear() {
    this._queries = [];
  }
}

module.exports = { BulkInserter, Pipeline, esc };

File: rust/src/bulk.rs
use std::collections::VecDeque;
use std::time::{Duration, Instant};

use crate::client::VedaClient;
use crate::error::VedaError;
use crate::protocol::Command;
use crate::result::{Value, VedaResult};

/// A pipeline of commands that can be sent in a single batch.
pub struct Pipeline {
    commands: Vec<PipelineCommand>,
}

#[derive(Debug, Clone)]
struct PipelineCommand {
    sql: String,
    params: Option<Vec<Value>>,
}

impl Pipeline {
    /// Create a new empty pipeline.
    pub fn new() -> Self {
        Pipeline {
            commands: Vec::new(),
        }
    }

    /// Add a query to the pipeline.
    pub fn query(mut self, sql: &str, params: Option<&[Value]>) -> Self {
        self.commands.push(PipelineCommand {
            sql: sql.to_string(),
            params: params.map(|p| p.to_vec()),
        });
        self
    }

    /// Add an execute to the pipeline.
    pub fn execute(mut self, sql: &str, params: Option<&[Value]>) -> Self {
        self.commands.push(PipelineCommand {
            sql: sql.to_string(),
            params: params.map(|p| p.to_vec()),
        });
        self
    }

    /// Add an insert statement.
    pub fn insert(mut self, table: &str, columns: &[&str], values: &[Value]) -> Self {
        let cols = columns.join(", ");
        let vals: Vec<String> = values.iter().map(|v| v.to_string()).collect();
        let sql = format!(
            "INSERT INTO {} ({}) VALUES ({});",
            table,
            cols,
            vals.join(", ")
        );
        self.commands.push(PipelineCommand { sql, params: None });
        self
    }

    /// Add an update statement.
    pub fn update(mut self, table: &str, set: &[(&str, Value)], where_clause: Option<&str>) -> Self {
        let set_clause: Vec<String> = set
            .iter()
            .map(|(col, val)| format!("{} = {}", col, val))
            .collect();
        let mut sql = format!("UPDATE {} SET {}", table, set_clause.join(", "));
        if let Some(w) = where_clause {
            sql.push_str(&format!(" WHERE {}", w));
        }
        sql.push(';');
        self.commands.push(PipelineCommand { sql, params: None });
        self
    }

    /// Add a delete statement.
    pub fn delete(mut self, table: &str, where_clause: Option<&str>) -> Self {
        let mut sql = format!("DELETE FROM {}", table);
        if let Some(w) = where_clause {
            sql.push_str(&format!(" WHERE {}", w));
        }
        sql.push(';');
        self.commands.push(PipelineCommand { sql, params: None });
        self
    }

    /// Number of commands in the pipeline.
    pub fn len(&self) -> usize {
        self.commands.len()
    }

    pub fn is_empty(&self) -> bool {
        self.commands.is_empty()
    }

    /// Execute all commands in the pipeline sequentially.
    pub fn execute_all(&self, client: &mut VedaClient) -> Result<Vec<VedaResult>, VedaError> {
        let mut results = Vec::with_capacity(self.commands.len());
        for cmd in &self.commands {
            let result = client.query(&cmd.sql, cmd.params.as_deref())?;
            results.push(result);
        }
        Ok(results)
    }

    /// Execute with a transaction wrapper.
    pub fn execute_transactional(
        &self,
        client: &mut VedaClient,
    ) -> Result<Vec<VedaResult>, VedaError> {
        client.transaction(|client| {
            let mut results = Vec::with_capacity(self.commands.len());
            for cmd in &self.commands {
                let result = client.query(&cmd.sql, cmd.params.as_deref())?;
                results.push(result);
            }
            Ok(results)
        })
    }

    /// Clear all commands.
    pub fn clear(&mut self) {
        self.commands.clear();
    }
}

impl Default for Pipeline {
    fn default() -> Self {
        Self::new()
    }
}

/// Bulk inserter that batches rows for efficient insertion.
pub struct BulkInserter {
    table: String,
    columns: Vec<String>,
    buffer: VecDeque<Vec<Value>>,
    batch_size: usize,
    total_inserted: u64,
    total_buffered: u64,
    last_flush: Instant,
    auto_flush_interval: Option<Duration>,
}

impl BulkInserter {
    /// Create a new bulk inserter.
    pub fn new(table: String, batch_size: usize) -> Self {
        BulkInserter {
            table,
            columns: Vec::new(),
            buffer: VecDeque::with_capacity(batch_size * 2),
            batch_size: batch_size.max(1),
            total_inserted: 0,
            total_buffered: 0,
            last_flush: Instant::now(),
            auto_flush_interval: None,
        }
    }

    /// Set column names for inserts.
    pub fn columns(mut self, columns: &[&str]) -> Self {
        self.columns = columns.iter().map(|s| s.to_string()).collect();
        self
    }

    /// Set auto-flush interval.
    pub fn auto_flush_interval(mut self, interval: Duration) -> Self {
        self.auto_flush_interval = Some(interval);
        self
    }

    /// Queue a row for insertion.
    pub fn queue(&mut self, row: Vec<Value>) -> Result<(), VedaError> {
        if !self.columns.is_empty() && row.len() != self.columns.len() {
            return Err(VedaError::Query(format!(
                "row has {} columns, expected {}",
                row.len(),
                self.columns.len()
            )));
        }
        self.buffer.push_back(row);
        self.total_buffered += 1;

        if self.buffer.len() >= self.batch_size {
            return Ok(()); // Caller should check and flush
        }

        // Check auto-flush
        if let Some(interval) = self.auto_flush_interval {
            if self.last_flush.elapsed() >= interval && !self.buffer.is_empty() {
                return Ok(()); // Signal to flush
            }
        }

        Ok(())
    }

    /// Queue a row from a map of column values.
    pub fn queue_map(&mut self, row: std::collections::HashMap<String, Value>) -> Result<(), VedaError> {
        if self.columns.is_empty() {
            return Err(VedaError::Query(
                "columns must be set before queueing maps".to_string(),
            ));
        }
        let values: Vec<Value> = self
            .columns
            .iter()
            .map(|col| row.get(col).cloned().unwrap_or(Value::Null))
            .collect();
        self.queue(values)
    }

    /// Flush the current buffer to the database.
    pub fn flush(&mut self, client: &mut VedaClient) -> Result<u64, VedaError> {
        if self.buffer.is_empty() {
            return Ok(0);
        }

        if self.columns.is_empty() {
            return Err(VedaError::Query(
                "columns must be set before flushing".to_string(),
            ));
        }

        let batch: Vec<Vec<Value>> = self.buffer.drain(..).collect();
        let count = batch.len() as u64;

        let cmd = Command::BulkInsert {
            table: self.table.clone(),
            columns: self.columns.clone(),
            rows: batch,
        };

        if let Some(protocol) = client.protocol() {
            let response = protocol.send_command(&cmd)?;
            match response.payload {
                crate::protocol::ResponsePayload::Ok(result) => {
                    self.total_inserted += result.row_count as u64;
                    self.last_flush = Instant::now();
                    Ok(count)
                }
                _ => Err(VedaError::Protocol(
                    "unexpected response for bulk insert".to_string(),
                )),
            }
        } else {
            Err(VedaError::Connection {
                message: "no protocol available".to_string(),
                host: None,
                port: None,
            })
        }
    }

    /// Flush if buffer is at or exceeds batch_size.
    pub fn flush_if_needed(&mut self, client: &mut VedaClient) -> Result<u64, VedaError> {
        if self.buffer.len() >= self.batch_size {
            self.flush(client)
        } else {
            Ok(0)
        }
    }

    /// Should the caller flush?
    pub fn should_flush(&self) -> bool {
        if self.buffer.len() >= self.batch_size {
            return true;
        }
        if let Some(interval) = self.auto_flush_interval {
            if self.last_flush.elapsed() >= interval && !self.buffer.is_empty() {
                return true;
            }
        }
        false
    }

    /// Get the number of rows in the buffer.
    pub fn buffered(&self) -> usize {
        self.buffer.len()
    }

    /// Get total rows inserted so far.
    pub fn total_inserted(&self) -> u64 {
        self.total_inserted
    }

    /// Get total rows queued (including unflushed).
    pub fn total_buffered(&self) -> u64 {
        self.total_buffered
    }

    /// Clear the buffer without inserting.
    pub fn clear(&mut self) {
        self.buffer.clear();
    }
}

/// Streaming bulk loader that reads from a source and inserts in batches.
pub struct StreamingBulkLoader {
    inserter: BulkInserter,
}

impl StreamingBulkLoader {
    pub fn new(table: String, columns: Vec<String>, batch_size: usize) -> Self {
        StreamingBulkLoader {
            inserter: BulkInserter::new(table, batch_size).columns(
                &columns.iter().map(|s| s.as_str()).collect::<Vec<_>>(),
            ),
        }
    }

    /// Load rows from an iterator.
    pub fn load_from_iterator<I>(
        &mut self,
        client: &mut VedaClient,
        rows: I,
    ) -> Result<u64, VedaError>
    where
        I: Iterator<Item = Vec<Value>>,
    {
        let mut total = 0u64;
        for row in rows {
            self.inserter.queue(row)?;
            if self.inserter.should_flush() {
                total += self.inserter.flush(client)?;
            }
        }
        // Final flush
        if self.inserter.buffered() > 0 {
            total += self.inserter.flush(client)?;
        }
        Ok(total)
    }

    /// Get the underlying inserter stats.
    pub fn stats(&self) -> (u64, u64) {
        (self.inserter.total_inserted(), self.inserter.total_buffered())
    }
}
